### ISSUE TITLE
HTMLElement#click() should not be limited to activatable elements.

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#6542]
The following should log `true`:
```js
var b = document.createElement("b");
var hit = false;
b.addEventListener("click", function() { hit = true; })
b.click()
console.log(hit)
```

(This will need to be converted into an automated wpt test.)

CC @Manishearth

**Comment by rebstar6** [comment]
dibs!

**Comment by jdm** [comment]
This is the `Click` method in `components/script/dom/htmlelement.rs`. We should extract the implementation of `synthetic_click_activation` from the `Activatable` trait in `components/script/dom/activation.rs`, and have it obtain an optional `Activatable` value when invoking the appropriate steps from https://html.spec.whatwg.org/multipage/interaction.html#run-synthetic-click-activation-steps (rather than gating the whole algorithm on obtaining an `Activatable` value as it currently does).

**Comment by Manishearth** [comment]
Actually, it should just take an `Element`, and when it reaches step 3 "Run pre-click activation steps on the element.", it should then do `elem.as_maybe_activatable().pre_click_activation()`. Ditto for the other steps.

**Comment by jdm** [comment]
@Manishearth Yes, that's what I tried to imply in my description.

**Comment by rebstar6** [comment]
Hmm, so for a 1st step I'd want to make the test that needs to be added. Can one of you point me to where this should go?

**Comment by Manishearth** [comment]
In /html/editing/activation/. There already is a click.html test there that; but the editing folder seems to be disabled?

**Comment by rebstar6** [comment]
Hmm, yeah - and if I un-disable, pretty much everything fails:
Ran 17 tests finished in 21.0 seconds.
  • 1 ran as expected. 0 tests skipped.
  • 1 tests timed out unexpectedly
  • 16 tests had unexpected subtest results

**Comment by rebstar6** [comment]
Also, can one of you verify this is right. Listing out the spec items, want to make sure I have straight what things should occur for all elements vs just some:

1. If the element's click in progress flag is set to true, then abort these steps. -- DO FOR ALL --
2. Set the click in progress flag on the element to true. -- DO FOR ALL --
3. Run pre-click activation steps on the element. -- ONLY IF ACTIVATIBLE --
4. Fire a click event at the element. -- DO FOR ALL --
_If the run synthetic click activation steps algorithm was invoked because the click() method was invoked, then the isTrusted attribute must be initialised to false._ -- not seeing this in the code anywhere? Is this another issue? --
5. If this click event is not canceled, run post-click activation steps on the element. If the event is canceled, the user agent must run canceled activation steps on the element instead. -- ONLY IF ACTIVATIBLE --
6. Set the click in progress flag on the element to false. -- DO FOR ALL --

Also, note the question about step 4 above - is the isTrusted piece elsewhere or just not implemented yet?

**Comment by Manishearth** [comment]
> Also, can one of you verify this is right. 

It is. The three activation steps [here](https://html.spec.whatwg.org/multipage/interaction.html#run-pre-click-activation-steps) are the only ones that are selectively invoked.

> not seeing this in the code anywhere? Is this another issue?

Yep. Straightforward to fix, add another parameter to `synthetic_click_activation` and use it.

**Comment by rebstar6** [comment]
OK, progress so far in https://github.com/rebstar6/servo/commit/8f8617f894de56e55b5ad18ce75b6f25d27aeaa4.
A few things I wanna confirm before move ahead with the isTrusted piece:
- I added a fromClick bool parameter to the synthetic_click_activation method. Setting to false for all callers but the click method in htmlelement. This will be used for setting isTrusted (?)
- Other callers (htmllabelelement, htmlinputelement, htmlbuttonelement, and document) were all checking for activatable before calling synthetic_click_activation. htmllabelelement and htmlbuttonelement are both referring to [implicit submission](https://html.spec.whatwg.org/multipage/forms.html#implicit-submission), which seems unclearly defined to me. Any idea if those should still do the pre-call activatable check?
- Also, for all of the above, are there tests I can validate with?
Thanks!

**Comment by jdm** [comment]
Good questions!
* Instead of a boolean, I would recommend a two-state enum type instead, since this ends up being much more self-describing when reading the code later than true/false.
* I think your changes are correct and the other callers should always invoke the synthetic activation.
* `./mach test-wpt tests/wpt/web-platform-tests/html/semantics/forms/` will get a bunch of ones that invoke `click()`, and there's `tests/wpt/web-platform-tests/html/editing/activation/click.html`, and there's a manual test in `tests/html/test-synthetic-click-activation.html`. There's nothing clearly testing labels, so that would be a good thing to verify.

In fact, it looks like HTMLLabelElement can get stuck in an endless loop of synthetic_click_activation, since that invokes activation_behaviour, which invokes synthetic_click_activation. We should write a test for this!

**Comment by Manishearth** [comment]
> In fact, it looks like HTMLLabelElement can get stuck in an endless loop of synthetic_click_activation, since that invokes activation_behaviour, which invokes synthetic_click_activation. 

Yeah, there should be a `GetControl()` in there


>  implicit submission, which seems unclearly defined to me. Any idea if those should still do the pre-call activatable check?

Yep. From the spec:

> then doing so for a form whose default button _has a defined activation behaviour_ must cause the user agent to run synthetic click activation steps on that default button.

**Comment by jdm** [comment]
@Manishearth Which part of the spec is that quote from?

**Comment by Manishearth** [comment]
@jdm https://html.spec.whatwg.org/multipage/forms.html#implicit-submission

**Comment by jdm** [comment]
Any further issues encountered, @rebstar6?

**Comment by rebstar6** [comment]
Nope, sorry, just was super busy last week with some other stuff - will be getting back to it this week.

**Comment by timvandermeij** [comment]
Fixed by #9930?

**Comment by jdm** [comment]
Yep!
